File: Leetcode/TypeScript/560. Subarray Sum Equals K.ts
function subarraySum(nums: number[], k: number): number {
    let m = new Map();
    m.set(0,1);
    let cur=0, ans=0;
    
    for(let i=0; i<nums.length; i++) {
        cur += nums[i];
        if(m.has(cur-k)) ans += m.get(cur-k);
        m.set(cur, (m.get(cur) ?? 0) + 1);
    }
    
    return ans;
};

File: Leetcode/C++/560. Subarray Sum Equals K.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> um;
        um[0]++;
        
        int cur = 0, ans = 0;
        for(auto x : nums) {
            cur += x;
            if(um.find(cur-k) != um.end()) {
                ans += um[cur-k];
            }
            um[cur]++;
        }
        
        return ans;
    }
};
